lektor/imagetools.py

```diff
@@ -440,40 +492,74 @@ def process_image(ctx, source_image, dst_filename, width, height=None,
     portable_popen(cmdline).wait()
 
 
-def make_thumbnail(ctx, source_image, source_url_path, width, height=None,
-                   crop=False, quality=None):
+def make_thumbnail(ctx, source_image, source_url_path,
+                   width=None, height=None, mode=THUMBNAIL_MODE.DEFAULT,
+                   upscale=None, quality=None):
     """Helper method that can create thumbnails from within the build process
     of an artifact.
     """
+    if width is None and height is None:
+        raise ValueError("Must specify at least one of width or height.")
+
+    if upscale is None:
+        upscale = {
+            THUMBNAIL_MODE.FIT: False,
+            THUMBNAIL_MODE.CROP: True,
+            THUMBNAIL_MODE.STRETCH: True,
+        }[mode]
+
+    # fallback to "fit" in case of erroneous arguments (but preserve
+    # the `upscale` defaults as defined above).
+    if mode != THUMBNAIL_MODE.FIT and (width is None or height is None):
+        warnings.warn(
+            '"%s" mode requires both `width` and `height` to be defined. '
+            'Falling back to "fit" mode.`' % mode.label
+        )
+        mode = THUMBNAIL_MODE.FIT
```

Comment:
Why not fail with a bang here? Isn't the user misusing the API?

Reply:
Yes, that was it. Rereading the last half of our talking on #540, we were talking about a similar issue but not quite the same thing.

To summarize it all, I think the rule-of-thumb I was going with, is that if you have a template query that can make sense for certain image dimensions, but not others, have a graceful fallback and throw a warning, but don't trip a build failure. For instance, this situation could happen if `upscale=False, mode=crop` and the image is too small to crop for given dimensions. If the query never makes sense though for any image, throw an exception and fail the build. This if block fits this situation, where it never makes sense to crop without given dimensions.

That said, I do not want to break projects that build successfully right now without a transition deprecation period. Currently `.thumbnail(500, crop=True)` builds successfully. I don't want to break that yet. Going with our revert-one-commit-later strategy though, we could throw an exception here, and then dial it back to a deprecation warning in the final commit.